src/daemon/plugin.h

```diff
@@ -243,21 +243,19 @@ int plugin_shutdown_all(void);
  *  plugin_write
  *
  * DESCRIPTION
- *  Calls the write function of the given plugin with the provided data set and
- *  value list. It differs from `plugin_dispatch_values' in that it does not
- *  update the cache, does not do threshold checking, call the chain subsystem
- *  and so on. It looks up the requested plugin and invokes the function, end
- *  of story.
+ *  Directly enqueues the given metric family without updating the cache or
+ *  calling the chain subsystem. The metric family may still be dropped from
+ *  the queue if collectd is stopped or the queue length surpasses the
+ *  configured limit.
  *
  * ARGUMENTS
  *  plugin     Name of the plugin. If NULL, the value is sent to all registered
  *             write functions.
- *  m   The actual value to be processed. Must not be NULL.
+ *  fam   The actual metric family to be processed.
  *
  * RETURN VALUE
- *  Returns zero upon success or non-zero if an error occurred. If `plugin' is
- *  NULL and more than one plugin is called, an error is only returned if *all*
- *  plugins fail.
+ *  Returns zero if the metric family was sucessfully enqueued or non-zero if
+ *  an error occurred.
```

Comment:
Do you have a write overflow test for both of the cases where plugin is specified, and where it is NULL (i.e. where family should go to all registered writers)?

Reply:
Having `plugin == NULL` is actually the more common case (e.g. what usually happens when dispatching a metric family) and why I decided to implement this PR using a single queue with one tail per writer. Otherwise one queue per write would likely have been more efficient.

As far as I can tell calls to `plugin_write` _with_ a plugin name could happen via:

- Filter Chains
- Custom Python / Perl plugins (they have bindings for `plugin_write`)
- Custom C plugins that call `plugin_write` directly instead of going the usual dispatch route

I did however not play around with any of those.

Reply:
Ah, right.  If filter chains are about aggregation, I do not think that has been ported to v6 API yet (see #4031).

If your earlier valgrind testing did not include "plugin != NULL" `plugin_write `case, could you check that just by using some specific plugin, e.g. `write_prometheus`?

IMHO `write_prometheus` is good test-case because MHD library used by it, also uses threads.  Testing would require also querying the write plugin, like this:
`http_proxy= curl --no-progress-meter http://127.0.0.1:<port>/metrics`